File: Decoder.java
import java.nio.charset.StandardCharsets;
import java.util.HashMap;
import java.util.Map;

public class Decoder {
    static long getIdx(char c) {
        long x = c;
        if (x >= 65) {
            return x - 65;
        }
        return x - 65 + 41;
    }

    static boolean checkVersion(String s) {
        long wi = getIdx(s.charAt(0)) * 32 + getIdx(s.charAt(1));
        long x = getIdx(s.charAt(2));
        long check = getIdx(s.charAt(3));

        return 1 >= (wi * x + check) % 32;
    }

    public static String decode(String enc) {
        if (!checkVersion(enc)) {
            return "";
        }
        Tree tree = new Tree();
        tree.init(enc.substring(0, 4));
        String rawEncode = tree.decode(enc);
        return new String(java.util.Base64.getDecoder().decode(rawEncode), StandardCharsets.UTF_8);
    }
}


class Node {
    Character chat = '.';
    Map<Character, Node> children;

    public Map<Character, Node> getChildren() {
        return children;
    }

    public Node() {
        children = new HashMap<>();
    }

    public Character getChar() {
        return chat;
    }

    public void setChar(Character charr) {
        this.chat = charr;
    }

    public void setChildren(Character k, Node v) {
        this.children.put(k, v);
    }
}

class Num {
    static long get(long num) {
        return (num >>> 0) % 4294967296L;
    }

    static long xor(long a, long b) {
        return get(get(a) ^ get(b));
    }

    static long and(long a, long b) {
        return get(get(a) & get(b));
    }

    public static long mul(long a, long b) {
        long high16 = ((a & 0xffff0000L) >>> 0) * b;
        long low16 = (a & 0x0000ffffL) * b;
        return get((high16 >>> 0) + (low16 >>> 0));
    }

    static long or(long a, long b) {
        return get(get(a) | get(b));
    }

    static long not(long a) {
        return get(~get(a));
    }

    static long shiftLeft(long a, long b) {
        return get(get(a) << b);
    }

    static long shiftRight(long a, long b) {
        return get(a) >>> b;
    }

    static long mod(long a, long b) {
        return get(get(a) % b);
    }
}

class Random {
    static final long MIN_LOOP = 8;
    static final long PRE_LOOP = 8;

    static final long BAY_SH0 = 1;
    static final long BAY_SH1 = 10;
    static final long BAY_SH8 = 8;
    static final long BAY_MASK = 0x7fffffff;
    long[] status = new long[4];
    long mat1;
    long mat2;
    long tmat;

    void seed(String seeds) {
        for (int i = 0; i < 4; i++) {
            if (seeds.length() > i) {
                this.status[i] = Num.get(seeds.charAt(i));
            } else {
                this.status[i] = Num.get(110);
            }
        }
        mat1 = status[1];
        mat2 = status[2];
        tmat = status[3];
        init();
    }

    void init() {
        for (int idx = 0; idx < MIN_LOOP - 1; idx++) {
            int idx1 = idx + 1;
            long status1 = this.status[(idx1) & 3];
            long status2 = this.status[idx & 3];
            long shiftedStatus = Num.shiftRight(status2, 30);
            long xorResult1 = Num.xor(status2, shiftedStatus);
            long mulResult = Num.mul(1812433253, xorResult1);
            long addResult = idx1 + mulResult;
            long xorResult2 = Num.xor(status1, addResult);
            this.status[idx1 & 3] = xorResult2;
        }
        if ((status[0] & BAY_MASK) == 0 && status[1] == 0 && status[2] == 0 && status[3] == 0) {
            status[0] = 66;
            status[1] = 65;
            status[2] = 89;
            status[3] = 83;
        }

        for (long idx = 0; idx < PRE_LOOP; idx++) {
            nextState();
        }
    }

    private void nextState() {
        long x;
        long y = status[3];
        x = Num.xor(Num.and(status[0], BAY_MASK), Num.xor(status[1], status[2]));
        x = Num.xor(x, Num.shiftLeft(x, BAY_SH0));
        y = Num.xor(y, Num.xor(Num.shiftRight(y, BAY_SH0), x));
        status[0] = status[1];
        status[1] = status[2];
        status[2] = Num.xor(x, Num.shiftLeft(y, BAY_SH1));
        status[3] = y;
        status[1] = Num.xor(status[1], Num.and(-Num.and(y, 1), mat1));
        status[2] = Num.xor(status[2], Num.and(-Num.and(y, 1), mat2));
    }

    public int generate(long max) {
        nextState();
        long t0 = status[3];
        long t1 = Num.xor(status[0], Num.shiftRight(status[2], BAY_SH8));
        t0 = Num.xor(t0, t1);
        t0 = Num.xor(Num.and(-Num.and(t1, 1), tmat), t0);
        return (int) (((t0 % max) + max) % max);
    }
}

class Tree {
    static final String B32_CODE = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567";
    static final String B64_CODE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/=";
    static final long[] CNT = {1, 2, 2, 2, 2, 2};
    Random random;
    String sign;
    Map<Character, String> inner;

    Node head;

    Tree() {
        this.random = new Random();
        sign = "";
        inner = new HashMap<>();
        head = new Node();
    }

    void init(String sign) {
        this.random.seed(sign);
        this.sign = sign;
        for (int i = 0; i <= 63; i++) {
            addSymbol(B64_CODE.charAt(i), CNT[(i + 1) / 11]);
        }
    }

    private String addSymbol(char ch, long len) {
        Node ptr = this.head;
        StringBuilder symbol = new StringBuilder();
        for (long i = 0; i < len; i++) {
            char innerChar = B32_CODE.charAt(random.generate(32));

            while (ptr.getChildren().containsKey(innerChar) && ptr.getChildren().get(innerChar).getChar() != '.') {
                innerChar = B32_CODE.charAt(random.generate(32));
            }
            symbol.append(innerChar);
            if (!ptr.getChildren().containsKey(innerChar)) {
                ptr.setChildren(innerChar, new Node());
            }
            ptr = ptr.getChildren().get(innerChar);
        }
        ptr.setChar(ch);
        inner.put(ch, symbol.toString());
        return symbol.toString();
    }

    String decode(String enc) {
        StringBuilder dec = new StringBuilder();
        for (int i = 4; i < enc.length(); ) {
            if (enc.charAt(i) == '=') {
                dec.append('=');
                i++;
                continue;
            }
            Node ptr = head;
            while (i < enc.length() && ptr.getChildren().containsKey(enc.charAt(i))) {
                ptr = ptr.getChildren().get(enc.charAt(i));
                i++;
            }
            dec.append(ptr.getChar());
        }
        return dec.toString();
    }
}

File: Decoder.py
import base64

# python版本的api.js

class Decoder:
    @staticmethod
    def get_idx(c):
        x = ord(c)
        if x >= 65:
            return x - 65
        return x - 65 + 41

    @staticmethod
    def check_version(s):
        wi = Decoder.get_idx(s[0]) * 32 + Decoder.get_idx(s[1])
        x = Decoder.get_idx(s[2])
        check = Decoder.get_idx(s[3])
        return 1 >= (wi * x + check) % 32

    @staticmethod
    def decode(enc):
        if not Decoder.check_version(enc):
            return ""
        tree = Tree()
        tree.init(enc[:4])
        raw_encode = tree.decode(enc)
        return base64.b64decode(raw_encode).decode('utf-8')


class Node:
    def __init__(self):
        self.chat = '.'
        self.children = {}

    def get_char(self):
        return self.chat

    def set_char(self, charr):
        self.chat = charr

    def set_children(self, k, v):
        self.children[k] = v

    def get_children(self):
        return self.children


class Num:
    @staticmethod
    def get(num):
        return num & 0xFFFFFFFF

    @staticmethod
    def xor(a, b):
        return Num.get(a ^ b)

    @staticmethod
    def and_(a, b):
        return Num.get(a & b)

    @staticmethod
    def mul(a, b):
        high16 = (a & 0xffff0000) * b
        low16 = (a & 0x0000ffff) * b
        return Num.get(high16 + low16)

    @staticmethod
    def or_(a, b):
        return Num.get(a | b)

    @staticmethod
    def not_(a):
        return Num.get(~a)

    @staticmethod
    def shift_left(a, b):
        return Num.get(a << b)

    @staticmethod
    def shift_right(a, b):
        return Num.get(a >> b)

    @staticmethod
    def mod(a, b):
        return Num.get(a % b)


class Random:
    MIN_LOOP = 8
    PRE_LOOP = 8
    BAY_SH0 = 1
    BAY_SH1 = 10
    BAY_SH8 = 8
    BAY_MASK = 0x7fffffff

    def __init__(self):
        self.status = [0] * 4
        self.mat1 = 0
        self.mat2 = 0
        self.tmat = 0

    def seed(self, seeds):
        for i in range(4):
            self.status[i] = Num.get(ord(seeds[i])) if i < len(seeds) else Num.get(110)
        self.mat1 = self.status[1]
        self.mat2 = self.status[2]
        self.tmat = self.status[3]
        self.init()

    def init(self):
        for idx in range(Random.MIN_LOOP - 1):
            idx1 = idx + 1
            status1 = self.status[idx1 & 3]
            status2 = self.status[idx & 3]
            shifted_status = Num.shift_right(status2, 30)
            xor_result1 = Num.xor(status2, shifted_status)
            mul_result = Num.mul(1812433253, xor_result1)
            add_result = idx1 + mul_result
            xor_result2 = Num.xor(status1, add_result)
            self.status[idx1 & 3] = xor_result2

        if all(s == 0 for s in self.status):
            self.status = [66, 65, 89, 83]

        for _ in range(Random.PRE_LOOP):
            self.next_state()

    def next_state(self):
        y = self.status[3]
        x = Num.xor(Num.and_(self.status[0], Random.BAY_MASK), Num.xor(self.status[1], self.status[2]))
        x = Num.xor(x, Num.shift_left(x, Random.BAY_SH0))
        y = Num.xor(y, Num.xor(Num.shift_right(y, Random.BAY_SH0), x))
        self.status = [self.status[1], self.status[2], Num.xor(x, Num.shift_left(y, Random.BAY_SH1)), y]
        self.status[1] = Num.xor(self.status[1], Num.and_(-Num.and_(y, 1), self.mat1))
        self.status[2] = Num.xor(self.status[2], Num.and_(-Num.and_(y, 1), self.mat2))

    def generate(self, max_val):
        self.next_state()
        t0 = self.status[3]
        t1 = Num.xor(self.status[0], Num.shift_right(self.status[2], Random.BAY_SH8))
        t0 = Num.xor(t0, t1)
        t0 = Num.xor(Num.and_(-Num.and_(t1, 1), self.tmat), t0)
        return (t0 % max_val + max_val) % max_val


class Tree:
    B32_CODE = "ABCDEFGHIJKLMNOPQRSTUVWXYZ234567"
    B64_CODE = "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/="
    CNT = [1, 2, 2, 2, 2, 2]

    def __init__(self):
        self.random = Random()
        self.sign = ""
        self.inner = {}
        self.head = Node()

    def init(self, sign):
        self.random.seed(sign)
        self.sign = sign
        for i in range(64):
            self.add_symbol(Tree.B64_CODE[i], Tree.CNT[(i + 1) // 11])

    def add_symbol(self, ch, length):
        ptr = self.head
        symbol = []
        for _ in range(length):
            inner_char = Tree.B32_CODE[self.random.generate(32)]
            while inner_char in ptr.get_children() and ptr.get_children()[inner_char].get_char() != '.':
                inner_char = Tree.B32_CODE[self.random.generate(32)]
            symbol.append(inner_char)
            if inner_char not in ptr.get_children():
                ptr.set_children(inner_char, Node())
            ptr = ptr.get_children()[inner_char]
        ptr.set_char(ch)
        self.inner[ch] = ''.join(symbol)
        return ''.join(symbol)

    def decode(self, enc):
        dec = []
        i = 4
        while i < len(enc):
            if enc[i] == '=':
                dec.append('=')
                i += 1
                continue
            ptr = self.head
            while i < len(enc) and enc[i] in ptr.get_children():
                ptr = ptr.get_children()[enc[i]]
                i += 1
            dec.append(ptr.get_char())
        return ''.join(dec)
